### PR TITLE
Bug 1796348 - Handle requesting notification permission for downloads

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1796348

### CODE REVIEW

**Comment by mcarare** [comment]
Notes: 

- this could have been simplified if the handling of the permission request was skipped, that could have been possible because we do not currently do something in case the permission was rejected, and the download proceeds as expected.

I do think we could do better than not showing the notification when permission is denied (showing in-app progress notification or similar). I also think we should reconsider this feature from a privacy POV. The download notification persists after you close the app, so the downloaded file name is visible even after the app and browsing history are cleared.

- the notification permission is not a mandatory one to perform the download, so it could not have been added to the required permissions handled in `DownloadsFeature` inside `onNeedToRequestPermissions`. Considering the fact that we need this permission only for displaying an active/ finished download, the permission request is shown after the download is confirmed, not before.

**Review by jonalmeida — COMMENTED** [review]
> I do think we could do better than not showing the notification when permission is denied (showing in-app progress notification or similar). I also think we should reconsider this feature from a privacy POV. 

This is a good idea, please file this feature request. It would make a lot of sense to do this if you are in the downloads UI to monitor the process.

> the notification permission is not a mandatory one to perform the download, so it could not have been added to the required permissions handled in DownloadsFeature inside onNeedToRequestPermissions... the permission request is shown after the download is confirmed, not before.

That's odd. Don't we want a download service that connects to a notification to ensure the app isn't killed until the download progress is done? Granted though that the `FetchDownloadManager` doesn't need this permission, only the `AbstractFetchDownloadService#performDownload` needs this.

I guess my question is, should we be requesting for notification permission in the same flow as the `DownloadsFeature#onNeedToRequestPermissions`? I agree that we can do better here in general.

**Comment by mcarare** [comment]
> > the notification permission is not a mandatory one to perform the download, so it could not have been added to the required permissions handled in DownloadsFeature inside onNeedToRequestPermissions... the permission request is shown after the download is confirmed, not before.
> 
> That's odd. Don't we want a download service that connects to a notification to ensure the app isn't killed until the download progress is done? Granted though that the `FetchDownloadManager` doesn't need this permission, only the `AbstractFetchDownloadService#performDownload` needs this.
> 
> I guess my question is, should we be requesting for notification permission in the same flow as the `DownloadsFeature#onNeedToRequestPermissions`? I agree that we can do better here in general.


We do have a service that ensures the app isn't killed, the service just does not have the permission to post notifications if this permission is denied.

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt [hunk 1]
```diff
@@ -79,6 +95,15 @@ class FetchDownloadManager<T : AbstractFetchDownloadService>(
         return download.id
     }
 
+    private fun onNotificationPermissionGranted() {
+        // this means we can show standard notifications
+    }
+
+    private fun onNotificationPermissionRejected() {
+        // this means we can show standard notifications
+        // maybe show in-app notifications?
```
[the remark this answers is not in the document]
Comment (mcarare) [reply android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt]: Opened https://bugzilla.mozilla.org/show_bug.cgi?id=1814863

**Review by jonalmeida — APPROVED** [review]
Speaking with mcarare offline, this seems like the best approach for now.

We need a much bigger change to our permission request model with Android 13, but it does not need us to block supporting the new API platform.

Thanks for pushing through these annoying changes! 💪

**Comment by mcarare** [comment]
Landing this after the beta cut.

**Comment by mcarare** [comment]
Closing in favor of https://github.com/mozilla-mobile/firefox-android/pull/849, which implements a centralized and common handling method for requesting notification permission and handling showing notifications.

**Pull request closed without merging**